crates/wasmparser/src/readers/core/reloc.rs

```diff
@@ -0,0 +1,307 @@
+use std::ops::Range;
+
+use crate::{BinaryReader, FromReader, Result, SectionLimited};
+
+/// Reader for relocation entries within a `reloc.*` section.
+pub type RelocationEntryReader<'a> = SectionLimited<'a, RelocationEntry>;
+
+/// Reader for reloc.* sections as defined by
+/// https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md#relocation-sections.
+#[derive(Debug, Clone)]
+pub struct RelocSectionReader<'a> {
+    section: u32,
+    range: Range<usize>,
+    entries: SectionLimited<'a, RelocationEntry>,
+}
+
+impl<'a> RelocSectionReader<'a> {
+    /// Creates a new reader for a `reloc.*` section starting at
+    /// `original_position` within the wasm file.
+    pub fn new(data: &'a [u8], original_position: usize) -> Result<Self> {
+        Self::from_reader(&mut BinaryReader::new_with_offset(data, original_position))
+    }
+
+    /// Index of section to which the relocations apply.
+    pub fn section_index(&self) -> u32 {
+        self.section
+    }
+
+    /// The byte range of the entire section.
+    pub fn range(&self) -> Range<usize> {
+        self.range.clone()
+    }
+
+    /// The relocation entries.
+    pub fn entries(&self) -> SectionLimited<'a, RelocationEntry> {
+        self.entries.clone()
+    }
+}
+
+impl<'a> FromReader<'a> for RelocSectionReader<'a> {
+    fn from_reader(reader: &mut BinaryReader<'a>) -> Result<Self> {
+        let range = reader.range().clone();
+        let section = reader.read_var_u32()?;
+        Ok(Self {
+            section,
+            range,
+            entries: SectionLimited::new(reader.remaining_buffer(), reader.original_position())?,
+        })
+    }
+}
+
+macro_rules! back_to_enum {
+    ($(#[$meta:meta])+ $vis:vis enum $name:ident {
+        $($(#[$vmeta:meta])* $vname:ident $(= $val:expr)?,)*
+    }) => {
+        $(#[$meta])*
+        $vis enum $name {
+            $($(#[$vmeta])* $vname $(= $val)?,)*
+        }
+
+        impl std::convert::TryFrom<u8> for $name {
+            type Error = ();
+
+            fn try_from(v: u8) -> Result<Self, Self::Error> {
+                match v {
+                    $(x if x == $name::$vname as u8 => Ok($name::$vname),)*
+                    _ => Err(()),
+                }
+            }
+        }
+    }
+}
+
+back_to_enum! {
+
+    /// Relocation entry type. Each entry type corresponds to one of the
+    /// `R_WASM_*` constants defined at
+    /// https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/BinaryFormat/WasmRelocs.def
+    /// and
+    /// https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md#relocation-sections.
+    #[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+    #[repr(u8)]
+    pub enum RelocationType {
```

Comment:
The number of relocations is probably going to change over time so I'm curious if you have thoughts on having a catch-all `Unknown(u32)` case? I think it's good idea to have named variants no matter what, but my thinking is that it might be reasonable to have a catch-all case for forwards-compatibility perhaps

Reply:
It isn't possible to parse an unknown relocation type because we need to know whether it has an addend.  Otherwise we might attempt to parse the first byte of the addend as the next relocation type.

Reply:
Good point yeah, sounds good!